### ISSUE TITLE
[FEATURE] draft eggs cli v-00, contextual awareness

### ISSUE THREAD

**Issue by Qu4k** [nestdotland/nest.land#186]
# eggs cli v-00, contextual awareness

adapted from [qu4k/drafts](https://github.com/Qu4k/drafts)

## abstract

remove the need of a configuration file for nest.land modules, gathering much
of the information from the module directory itself.

## current implementation

as it stands the basic - but complete - implementation of a nest.land eggs.json
file is:

```json
{
    "name": "module-name",
    "description": "Your brief module description",
    "version": "0.0.1",
    "entry": "./src/main.ts",
    "stable": true,
    "unlisted": false,
    "fmt": true,
    "repository": "https://github.com/your_name/your_project",
    "files": [
        "./mod.ts",
        "./src/**/*",
        "./README.md"
    ]
}
```

information for all the fields can be retrieved from the documentation:
<https://nest.land/docs#configuration>

## future implementation

the methods proposed are considered a direct substitute for the current
`eggs.json` file. this does not mean that `eggs.json` will be deprecated
though, since the user can always provide one. our focus should be ease
of use without the *need* of a file. if the user chooses to make use of
a file for fine control he is free to do so.

if no `eggs.json` is provided, the `publish` command will now consist of
a series of questions with default answers gathered from the methods
below. not all fields are required in every publish event.

### ... first of all

* the CLI must be able to call the API and gather info about the module
  after the first publish.
* "only first publish" indicates that this field is no longer needed after
  the first publish, but (with the exception of name) can be changed later with
  cli flag (eg. `--override-description`)
* NVC stands for "no version control".

### `name`

* only first publish
* suggestion can be retrieved from root directory name or `README.md`

### `description`

* only fist publish
* suggestion can be retrieved from `README.md` subtitle / line after title

### `version`

* every publish
* SVN:  `$ svn log <repo>/tags --limit 1`
* GIT:  `$ git describe`
* NVC:  the `CHANGELOG.md` file or CLI input

this information can be also retrieved without using `Deno.run()` by simply
parsing version control files (eg `.git/refs/tags/`).
this information can be also retrieved in a `CHANGELOG.md` file.

### `stable`

* arbitrary
* can be moved to cli input or as a `--unstable` flag.

### `fmt`

* arbitrary
* can be moved to cli input or as a `--fmt` flag.
  (eg: `would you like to run "deno fmt" before publishing?`).

### `repository`

* only first publish
* SVN:  `$ svn info`
* GIT:  `$ git remote -v`
* NVC:  CLI input

### `files`

files can be retrieved in an opt-out fashion. The CLI should be able to parse
a `.ignore`-type file, if provided, and exclude files/directories
from publishing:

* SVN:  `$ svn proplist -v -R dir/`
* GIT:  the `.gitignore` file
* NVC:  the `.eggsignore` file

**Comment by martonlederer** [comment]
I like this idea, sounds great and easy to use. But what if someone else is working on the same module too? (In the future, nest.land will allow contributors for modules)

**Comment by notfilippo** [comment]
Since context-gathered fields are only suggestion given on publish, and are determined by the a predicable set of deterministic rules you can assure compatibility in the same environment shared between two or more machines.

**Comment by t8** [comment]
I like this idea as well. I do agree that we'll need to stress that this is completely optional, but I think it would greatly improve the user experience for publishing modules.

This idea could be the beginning of a new "intelligent" system that procedurally gathers the data necessary to fill the configuration (but still user correctable of course).

I see this outline as the beginning of an algorithm quite honestly. 

No matter what we do, we cannot interrupt the core values of the platform in order to implement this system.

**Comment by littledivy** [comment]
No matter what people think about configuration, they are necessary...

We could do:
```javascript
if(pathExists("eggs.json")) {
  // Do config parsing
} else {
 // Do what qu4k said
}
```
That is, it could be completely optional...
But the current situation is pretty perfect cause we have both json and yaml configs to make sure people don't think of them as another package.json. Though, the idea is great!

**Comment by zorbyte** [comment]
We made a rule of thumb that we shouldn't interact with version control, so we can't use it.

**Comment by zorbyte** [comment]
We can use it as a recommendation though for the repo url

**Comment by t8** [comment]
@zorbyte I don't see the issue of working with version control if it is a completely optional feature and not required in any way. The fact that we allow users to specify their module repository is the exact same type of thing.

**Comment by zorbyte** [comment]
Oh I see. I want to make Version Control secondary, as in strictly a form of intelligent recommendations, otherwise users will be confused about the source of some of their configuration data.

**Comment by littledivy** [comment]
The landing page clearly says "Works without Git". Even if this will be optional...we should avoid having external dependencies not in our control

**Comment by t8** [comment]
@divy-work what things would you consider "external dependencies"?

**Comment by littledivy** [comment]
@tbaumer22 Like getting the repo using `git remote -v`, getting version either from Changelogs or SVN..etc are the external dependencies here.

**Comment by zorbyte** [comment]
3rd party code or data that allows any aspect of our service to function in any manner, critical or not.

**Comment by t8** [comment]
I think I'm following you now. Regardless of its importance, do you (@divy-work) and @zorbyte agree that this would be a beneficial **optional** feature?

**Comment by littledivy** [comment]
Yes, it can be a beneficial optional feature..but beware that when more users start using it..the config file will be ignored and people who don't even use git for projects will be unknown of the config file...but that's for the future, it think we can have it now...

**Comment by t8** [comment]
I think that as long as we're very specific in our documentation and CLI about how this system works, we should be able to safely navigate the situation.

**Comment by notfilippo** [comment]
> Oh I see. I want to make Version Control secondary, as in strictly a form of intelligent recommendations, otherwise users will be confused about the source of some of their configuration data.

The goal with this is to make it _possible_ not to use a configuration file. You can still provide one to override all the settings, but ideally you could just run `eggs publish` and <enter><enter><enter> so that cli prompts are filled with the default values (like many cli operate) and you are set without the _need_ of a file.

**Comment by notfilippo** [comment]
> 3rd party code or data that allows any aspect of our service to function in any manner, critical or not.

As it’s optional it’s a matter of:

- Can I run `this` to gather some context-aware data?
- Lets run `this` and see.

If it crashes, returns errors or simply is not present we can go to next section.

**Comment by notfilippo** [comment]
> The landing page clearly says "Works without Git". Even if this will be optional...we should avoid having external dependencies not in our control

Yeah, it should work without Git, but what stops me from hooking to it optionally, and only if there is a `.git/folder`, to gather intelligence in the project?

**Comment by notfilippo** [comment]
> No matter what people think about configuration, they are necessary...
> 
> 
> 
> We could do:
> 
> ```javascript
> 
> if(pathExists("eggs.json")) {
> 
>   // Do config parsing
> 
> } else {
> 
>  // Do what qu4k said
> 
> }
> 
> ```
> 
> That is, it could be completely optional...

“if no eggs.json is provided, the publish command will now consist of
a series of questions with default answers gathered from the methods
below. not all fields are required in every publish event.”

From the proposal.

> But the current situation is pretty perfect cause we have both json and yaml configs to make sure people don't think of them as another package.json. Though, the idea is great!

I don’t think providing the YAML format makes eggs.json/yaml different in any way from package.json. The real problem is **having a mandatory** configuration file. As Ryan reported in one of his talks, package.json is completely futile as you are just redeclaring things that can already be found in the actual project. 

The aim of this proposal is to change the default way that things are done with `eggs publish`, while providing backwards compatibility and leaving open an override option if a file is actually present in the root dir.

**Comment by notfilippo** [comment]
Also, nestdotland/nest.land#91 is related.
